src/limitless/limitless_monitor_service.cc

```diff
@@ -156,24 +157,66 @@ void LimitlessMonitorService::DecrementReferenceCounter(const std::string& servi
 }
 
 std::shared_ptr<HostInfo> LimitlessMonitorService::GetHostInfo(const std::string& service_id) {
-    std::lock_guard<std::mutex> services_guard(*(this->services_mutex));
-    if (!this->services.contains(service_id)) {
-        LOG(ERROR) << "Attempted to get host info for non-existent monitor with service ID " << service_id;
-        return nullptr;
+    std::vector<HostInfo> hosts;
+    SQLSTR connection_string;
+
+    {
+        std::lock_guard<std::mutex> services_guard(*(this->services_mutex));
+        if (!this->services.contains(service_id)) {
+            LOG(ERROR) << "Attempted to lock and get hosts for non-existent monitor with service ID " << service_id;
+            return nullptr;
+        }
+
+        std::shared_ptr<LimitlessMonitor> service = this->services[service_id];
+        std::lock_guard<std::mutex> limitless_routers_guard(*(service->limitless_routers_mutex));
+
+        if (service->limitless_routers == nullptr || service->limitless_routers->empty())
+            return nullptr;
+
+        // copy hosts
+        hosts = *(service->limitless_routers);
+        connection_string = service->limitless_router_monitor->GetConnectionString();
     }
 
-    std::shared_ptr<LimitlessMonitor> service = this->services[service_id];
+    std::unordered_map<std::string, std::string> properties;
 
-    std::lock_guard<std::mutex> limitless_routers_guard(*(service->limitless_routers_mutex));
-    std::vector<HostInfo> hosts = *(service->limitless_routers);
-    if (hosts.empty()) {
+    try {
+        RoundRobinHostSelector::SetRoundRobinWeight(hosts, properties);
+        HostInfo host = this->round_robin.GetHost(hosts, true, properties);
+        if (this->odbc_wrapper->TestConnectionToServer(connection_string, host.GetHost())) {
+            // the round robin host successfully connected
+            return std::make_shared<HostInfo>(host);
+        }
+    } catch (std::runtime_error& error) {
+        // no hosts available - just return nullptr
         return nullptr;
     }
```

Comment:
Not all runtime_errors are "no hosts available". This is why ideally it has its own error message. 

For now, just have it continue to Highest Weight.